lib/srv/discovery/fetchers/aws-sync/reconcile.go

```diff
@@ -24,640 +24,271 @@ import (
 	"google.golang.org/protobuf/proto"
 	"google.golang.org/protobuf/reflect/protoreflect"
 
-	accessgraphv1alpha "github.com/gravitational/teleport/gen/proto/go/accessgraph/v1alpha"
+	tag "github.com/gravitational/teleport/gen/proto/go/accessgraph/v1alpha"
 )
 
+func newResourceList() *tag.AWSResourceList {
+	return &tag.AWSResourceList{
+		Resources: make([]*tag.AWSResource, 0),
+	}
+}
+
 // ReconcileResults reconciles two Resources objects and returns the operations
 // required to reconcile them into the new state.
 // It returns two AWSResourceList objects, one for resources to upsert and one
 // for resources to delete.
-func ReconcileResults(old *Resources, new *Resources) (upsert, delete *accessgraphv1alpha.AWSResourceList) {
-	upsert, delete = &accessgraphv1alpha.AWSResourceList{}, &accessgraphv1alpha.AWSResourceList{}
-
-	for _, results := range []*reconcileIntermeditateResult{
-		reconcileUsers(old.Users, new.Users),
-		reconcileUserInlinePolicies(old.UserInlinePolicies, new.UserInlinePolicies),
-		reconcileUserAttachedPolicies(old.UserAttachedPolicies, new.UserAttachedPolicies),
-		reconcileUserGroups(old.UserGroups, new.UserGroups),
-		reconcileGroups(old.Groups, new.Groups),
-		reconcileGroupInlinePolicies(old.GroupInlinePolicies, new.GroupInlinePolicies),
-		reconcileGroupAttachedPolicies(old.GroupAttachedPolicies, new.GroupAttachedPolicies),
-		reconcilePolicies(old.Policies, new.Policies),
-		reconcileInstances(old.Instances, new.Instances),
-		reconcileS3(old.S3Buckets, new.S3Buckets),
-		reconcileRoles(old.Roles, new.Roles),
-		reconcileRoleInlinePolicies(old.RoleInlinePolicies, new.RoleInlinePolicies),
-		reconcileRoleAttachedPolicies(old.RoleAttachedPolicies, new.RoleAttachedPolicies),
-		reconcileInstanceProfiles(old.InstanceProfiles, new.InstanceProfiles),
-		reconcileEKSClusters(old.EKSClusters, new.EKSClusters),
-		reconcileAssociatedAccessPolicy(old.AssociatedAccessPolicies, new.AssociatedAccessPolicies),
-		reconcileAccessEntry(old.AccessEntries, new.AccessEntries),
-		reconcileAWSRDS(old.RDSDatabases, new.RDSDatabases),
-		reconcileSAMLProviders(old.SAMLProviders, new.SAMLProviders),
-		reconcileOIDCProviders(old.OIDCProviders, new.OIDCProviders),
-	} {
-		upsert.Resources = append(upsert.Resources, results.upsert.Resources...)
-		delete.Resources = append(delete.Resources, results.delete.Resources...)
+func ReconcileResults(old *Resources, new *Resources) (upsert, delete *tag.AWSResourceList) {
+	upsert, delete = newResourceList(), newResourceList()
+	reconciledResources := []*reconcilePair{
+		reconcile(old.Users, new.Users, usersKey, usersWrap),
```

Comment:
Since you're refactoring it anyway another improvement would be to create an interface and implement it for each type instead of passing multiple process function here.
That being said, that's probably something for a different PR.

Reply:
Agree, that could definitely be a different PR.